---
title: "DGE"
author: "Prasanth Sivakumar"
date: '`r Sys.Date()`'
output: html_document
---

```{r setup, include=FALSE}
require('knitr')
opts_chunk$set(echo = FALSE, warning = FALSE)
opts_knit$set(root.dir = 'C:/Users/Prasanth/Documents/ramsey/')
```

This is a summary report outlining the [**Gene Set Variation Analysis**](https://bioconductor.org/packages/devel/bioc/vignettes/GSVA/inst/doc/GSVA.html#1_Quick_start) of samples from the PRUMEC project. 40 Samples in total were used in the analysis, with an input counts matrix for 58,125 genes. 

Input data file names: 

* All samples raw counts matrix: 'Partek_prumec_skin_mRNA_HTSeq_Gene_counts_amended_IDs_26.10.2021.txt' \
* All samples metadata: 'prumec_metadata_final_305ID_amended_26.10.2021.xlsx' \
* Specific samples to use: 'Prasanth_IDs_P-LvsC.csv' \
* Gene sets: http://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp?collection=IMMUNESIGDB



```{r load_libraries, include=FALSE}
library(tidyverse)
library(DESeq2)
library(pheatmap)
library(ggrepel)
```

```{r load_data}
load('ramsey_s303_norm_counts_matrix.RData')
load('ramsey_s303_non_zero_counts_matrix.RData')
md <- read.csv('s303_metadata.csv')
```

```{r ids}
ids <- read.csv('C:/Users/Prasanth/google_drive/guys/resources/biomart_unique_ensembl_genes.csv')
```

```{r clean_data}
md <- md[md$sample_id != 'NAT 4', ]
# works after removing
md$matrix_id <- stringr::str_sort(colnames(zcm), numeric = T)
# check to make sure
#View(md[, c('sample_id', 'matrix_id')])

# re-sort whole md to match matrix variable order
md <- arrange(md, factor(matrix_id, levels = colnames(zcm)))
# convert matrix ids to row names
md <- column_to_rownames(md, 'matrix_id')

# edit md 'reactive lymph nodes' to shorten
md$sample_group <- gsub('Lymph Nodes\n$', 'LN', md$sample_group)
md$sample_group <- gsub('\n$', '', md$sample_group)
md$sample_group <- gsub(' ', '_', md$sample_group)
```

```{r filter_data}
smpls <- colnames(zcm)[grep('NAT', colnames(zcm))]
nat.cm <- zcm[, smpls]
nat.ncm <- ncm[, smpls]
nat.md <- md[smpls, ]

# sample groups
smpl.grps <- list(cll_ln = rownames(nat.md[nat.md$sample_group == 'CLL LN', ]), 
                  cll_pb = rownames(nat.md[nat.md$sample_group == 'CLL PB', ]), 
                  ctrl = rownames(nat.md[nat.md$sample_group == 'Control', ]), 
                  reac = rownames(nat.md[nat.md$sample_group == 'Reactive LN', ]))

# subset metadata by comparisons
mds <- list(ctrl_v_cll_ln = md[c(smpl.grps$ctrl, smpl.grps$cll_ln), ], 
            ctrl_v_cll_pb = md[c(smpl.grps$ctrl, smpl.grps$cll_pb), ], 
            ctrl_v_reac = md[c(smpl.grps$ctrl, smpl.grps$reac), ])

mds$ctrl_v_cll_ln$sample_group <- factor(mds$ctrl_v_cll_ln$sample_group, levels = c('Control', 'CLL LN'))
mds$ctrl_v_cll_pb$sample_group <- factor(mds$ctrl_v_cll_pb$sample_group, levels = c('Control', 'CLL PB'))
mds$ctrl_v_reac$sample_group <- factor(mds$ctrl_v_reac$sample_group, levels = c('Control', 'Reactive LN'))
```

Sample IDs of samples used in GSVA analysis

```{r modify_raw}
length(nat.md$sample_id)
nat.md$sample_id
```

Numbers of each sample type included in analysis

``` {r by_type}
table(nat.md$sample_group)
```

Size of filtered counts matrix for relevant samples and genes with no 0 counts

```{r filt_cm}
dim(nat.cm)
```

DESeq

```{r deseq}
#deseq.dataset <- DESeqDataSetFromMatrix(countData = nat.cm, 
deseq.dataset <- DESeqDataSetFromMatrix(countData =nat.cm[apply(nat.cm, 1, function(x) all(x !=0 )), ],
                       colData = nat.md, 
                       design = ~ sample_group)
```

```{r filt_dds}
deseq.dataset.filtered <- deseq.dataset[rowSums(counts(deseq.dataset)) >= 20]
deseq.dataset.filtered
```
## Differential gene expression

To analyse the differential expression across all genes in the dataset by the provided design, three steps are required, and all are run combined using a single function in DESeq2.

1. Normalisation
1. Estimated dispersion
1. Statistical analysis

These steps are sequentially run using the single _DESeq_ function, using the information already incorporated into the dds object.

```{r deseq}
deseq.analysis <- DESeq(deseq.dataset.filtered)
deseq.analysis
```


```{r res}
deseq.results <- list()
deseq.results$reactive_ln <- results(deseq.analysis, contrast = c('sample_group', 'Reactive_LN', 'Control'))
summary(deseq.results$reactive_ln, alpha = 0.05)
```

```{r res}
deseq.results$cll_pb <- results(deseq.analysis, contrast = c('sample_group', 'CLL_PB', 'Control'))
summary(deseq.results$cll_pb, alpha = 0.05)
```

```{r res}
deseq.results$cll_ln <- results(deseq.analysis, contrast = c('sample_group', 'CLL_LN', 'Control'))
summary(deseq.results$cll_ln, alpha = 0.05)
```
To explore the distribution of the differential gene expression results for both effect size (fold change) and confidence (p-value), a volcano plot is typically used. This can provide an overall pattern of differential gene expression in the dataset, split by upregulation and downregulation events. 

```{r results_df}
toDf <- function(res){
  arrange(as.data.frame(res), pvalue) %>% 
    rownames_to_column(var = 'ensembl_gene_id') %>%
    left_join(ids, by = 'ensembl_gene_id')
}

deseq.dfs <- lapply(deseq.results, toDf)
```




```{r volcano, warning=FALSE}
gfill <- c('Strongly up' = 'red', 'Slightly up' = 'pink', 
          'Strongly down' = 'blue', 'Slightly down' = 'lightblue', 
          'None' = 'grey')
gsize <- c('Strongly up' = 2, 'Slightly up' = 2, 
          'Strongly down' = 2, 'Slightly down' = 2, 
          'None' = 1)


plotVolcano <- function(res){
  as.data.frame(res) %>%
    filter(!is.na(padj)) %>%
    mutate(change = ifelse(padj < 0.05, 
                           ifelse(log2FoldChange > 0, 
                                  ifelse(log2FoldChange > log2(1.5), 'Strongly up', 'Slightly up'), 
                                  ifelse(log2FoldChange < -log2(1.5), 'Strongly down', 'Slightly down')), 
                           'None')) %>%
  ggplot(aes(log2FoldChange, -log10(pvalue))) + 
    geom_point(aes(fill = change, size = change), shape = 21, alpha = 0.5) +
    scale_fill_manual(values = gfill) +
    scale_size_manual(values = gsize) +
    geom_hline(yintercept = -log10(0.00265), linetype = 'dashed', alpha = 0.5) +
    geom_vline(xintercept = log2(1.5), linetype = 'dashed', alpha = 0.5) +
    geom_vline(xintercept = -log2(1.5), linetype = 'dashed', alpha = 0.5) +
    theme_classic()
}

plotVolcano(deseq.results$reactive_ln)
```

```{r volcano, warning=FALSE}
plotVolcano(deseq.results$cll_pb)
```

```{r volcano, warning=FALSE}
plotVolcano(deseq.results$cll_ln)
```

```{r clust}
# clusterSamples <- function(dds, svcm){
#   select <- order(rowMeans(counts(dds, normalized=TRUE)),
#                   decreasing=TRUE)[1:20]
#   df <- as.data.frame(colData(dds)[,'sample_group'])
#   pheatmap(assay(normTransform(dds))[select,], cluster_rows=FALSE, show_rownames=FALSE, cluster_cols=FALSE, annotation_col=df)
# }

#clusterSamples(deseq.analysis)
select <- order(rowMeans(counts(deseq.analysis, normalized=T)),
                  decreasing=T)[1:10]
df <- as.data.frame(colData(deseq.analysis)[, 'sample_group'], row.names = row.names(colData(deseq.analysis))) %>% `names<-`('Group')
pheatmap(assay(normTransform(deseq.analysis))[select,], cluster_rows=F, show_rownames=T, cluster_cols=T, annotation_col = df, scale = 'row')
```

```{r fh}
gselect <- order(rowMeans(counts(deseq.analysis, normalized=T)),
                  decreasing=T)
#gselect <- filter(deseq.dfs$reactive_ln, padj < 0.05)$ensembl_gene_id
df <- as.data.frame(colData(deseq.analysis)[, 'sample_group'], row.names = row.names(colData(deseq.analysis))) %>% `names<-`('Group')
pheatmap(assay(varianceStabilizingTransformation(deseq.analysis))[gselect, ], cluster_rows=F, show_rownames = F, cluster_cols = T, annotation_col = df, scale = 'row')

```

```{r cor}
vsd <- varianceStabilizingTransformation(deseq.analysis)
sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$sample_group, vsd$sample_id, sep="-")
colnames(sampleDistMatrix) <- NULL
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists)
```